drivers/scsi/mpt3sas/mpt3sas_ctl.c

```diff
@@ -4187,3 +4198,309 @@ mpt3sas_ctl_exit(ushort hbas_to_enumerate)
 	if (hbas_to_enumerate != 2)
 		misc_deregister(&gen2_ctl_dev);
 }
+
+/**
+ * mpt3_nvme_submit_command - submit NVME encapsulated command
+ * @ioc: per adapter object
+ * @cmd: NVME Command
+ * @ubuf: user buffer
+ * @kbuf: kernel Buffer
+ * @buf_len: buffer Length
+ * @result: cqe result
+ * @is_admin: admin or IO command
+ * @handle: to distinguish target NVME device
+ * @port_num: port number
+ */
+static int mpt3_nvme_submit_command(struct MPT3SAS_ADAPTER *ioc,
+	struct nvme_command cmd, u64 ubuf, void *kbuf, u32 buf_len,
+	u64 *result, int is_admin, u16 handle, u8 port_num)
+{
+	struct mpt3_nvme_kencap kencap;
+	struct mpt3_ioctl_command karg;
+	struct nvme_completion cqe;
+	int ret;
+	Mpi26NVMeEncapsulatedRequest_t *nvme_encap_request = kzalloc(sizeof(cmd) + \
+	    sizeof(Mpi26NVMeEncapsulatedRequest_t), GFP_KERNEL);
+
+	if (!nvme_encap_request)
+		return (-ENOMEM);
+
+	memset(&karg, 0, sizeof(karg));
+	memset(&cqe, 0, sizeof(cqe));
+	memset(&kencap, 0, sizeof(kencap));
+	karg.hdr.port_number = port_num;
+	karg.timeout = 120;
+
+	/*
+	 * data_sge_offset refers to the size of both NVME Encpasulated request
+	 * header plus the NVME command in words. data_sge_offset name is a bit
+	 * misleading for NVME Encpasulated request since PRP buffers will be
+	 * allocated by DMA.
+	 */
+	karg.data_sge_offset = (sizeof(cmd) + sizeof(*nvme_encap_request)) / 4;
+	karg.max_reply_bytes = 0;
+	karg.max_sense_bytes = sizeof(cqe);
+
+	/*
+	 * We can set reply_buf to MPI26_NVME_FLAGS_READ but not required
+	 */
+	kencap.reply_buf = NULL;
+
+	/*
+	 * Firmware writes CQE to sense_buf
+	 */
+	kencap.sense_buf =  &cqe;
+	kencap.nvme_encap_rqst = nvme_encap_request;
+	nvme_encap_request->Function = MPI2_FUNCTION_NVME_ENCAPSULATED;
+	nvme_encap_request->DataLength = buf_len;
+	nvme_encap_request->EncapsulatedCommandLength = sizeof(cmd);
+	nvme_encap_request->DevHandle = handle;
+	memcpy((u8*) nvme_encap_request + offsetof(Mpi26NVMeEncapsulatedRequest_t,
+	    NVMe_Command), &cmd, sizeof(cmd));
+
+	if (is_admin) {
+		nvme_encap_request->Flags = MPI26_NVME_FLAGS_SUBMISSIONQ_ADMIN | \
+		    MPI26_NVME_FLAGS_FORCE_ADMIN_ERR_RESP;
+	}
+
+	if (buf_len && (ubuf || kbuf)) {
+		/*
+		* Write opcode
+		*/
+		if (cmd.common.opcode & 0x1) {
+			if (ubuf)
+				karg.data_out_buf_ptr = (void __user *) ubuf;
+			else
+				kencap.dout_buf = kbuf;
+			karg.data_out_size = buf_len;
+			nvme_encap_request->Flags |= MPI26_NVME_FLAGS_WRITE;
+		}
+
+		/*
+		* Read opcode
+		*/
```

Comment:
Too much honor for two words comment.  I'd make it one line and may be on the same line as the `if`.  Or if multi-line comment, then it seems to miss spaces on following lines.

Reply:
Thanks! I made it a single-line comment.